irobotclient/main.py

```diff
@@ -0,0 +1,100 @@
+WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+Public License for more details.
+
+You should have received a copy of the GNU General Public License along
+with this program. If not, see <http://www.gnu.org/licenses/>.
+"""
+import errno
+
+from os import path
+from requests import Response
+
+from irobotclient import configuration_handler
+from irobotclient import request_formatter
+from irobotclient.custom_exceptions import IrobotClientException
+from irobotclient.request_handler import Requester
+
+
+def _print_error_details(error: OSError):
+    """
+    Print the output of any errors raised.
+
+    Utilises the OSError exception class which the iRobot Client uses for custom exceptions too.
+
+    :param error:
+    :return:
+    """
+    print("Inside _output_error_details") # Beth - debug
+    print(error)
+    exit(error.errno)
+
+
+def _download_data(response: Response, output_dir:str):
+    """
+
+    :param response:
+    :param output_dir:
+    :return:
+    """
+    try:
+        file_name = (path.split(response.url))[1]
+
+        if response.headers["content-type"] == "application/octet-stream":
+            with open(f"{output_dir}{file_name}", "wb") as file:
+                file.write(bytes(response.content))
+        else:
+            with open(f"{output_dir}{file_name}", "w") as file:
+                file.write(response.content)
+    except OSError as err:
+        _print_error_details(err)
+    except:
+        raise IrobotClientException(errno=errno.ECONNABORTED, message="Cannot write content to file.")
+
+
+def _run():
+    try:
+
+        # Set configurations from command line and/or environment.
+        config_details = configuration_handler.run()
+
+        file_extensions = request_formatter.get_file_extensions(config_details.input_file, config_details.no_index)
+        url = request_formatter.get_url_request_path(config_details.url, config_details.input_file)
+        headers = request_formatter.get_header(config_details.token)
+
+        if not file_extensions:
+            request_handler = Requester(url, headers)
+            response = request_handler.get_data()
+            _download_data(response, config_details.output_dir)
+            # TODO - checksum test if possible
+        else:
+            for ext in file_extensions:
+                # TODO - handle index file issues whereby a bam bai file may not be present but a pbi might.
+                request_handler = Requester(url + ext, headers)
+                response = request_handler.get_data()
+                _download_data(response, config_details.output_dir)
+                # TODO - checksum test for each file if possible
+
+        print("Exiting....")
+        exit()
+    except IrobotClientException as err:
+        _print_error_details(err)
+    except OSError as err:
+        _print_error_details(err)
+    except Exception as err:
```

Comment:
As with all things engineering: It depends™. Stack traces are useful for debugging, but not particularly friendly for end users. It's a judgement call as to who your users will be and what you think they'll find more useful. Also, some exceptions are obvious and won't necessarily need a stack trace (e.g., file not found).

That only applies to exceptions that you let bubble to the surface, rather than any you catch and handle earlier down the stack. As such, for now, I would err on the side of stack traces for this class of exceptions. In time, you'll learn what "normal behaviour" is; i.e., under normal operating conditions, what exceptions usually bubble to the surface. These will either be bugs or things you can deal with/explain to the user without bombarding them with a stack trace. Then, all that's left will be "weird behaviour", that will need a stack trace to be understood.